### PR TITLE
Packagist vpre support.

### PR DESCRIPTION
**Pull request by WouterDS**
This commit adds support for packagist vpre. #319

For example:

- stable symfony/symfony: stable symfony framework, latest dev: latest dev symfony framework
- stable laravel/framework: stable laravel framework, latest dev: latest laravel framework

### CODE REVIEW

File: server.js [hunk 1]
```diff
@@ -765,12 +766,44 @@ cache(function(data, match, sendBadge, request) {
...
+      var stableVersion = latestVersion(versions);
+      var vdata = versionColor(stableVersion);
+
+      var badgeText = null;
+      var badgeColor = null;
+
+      switch (info) {
+          case 'v':
+              badgeText = vdata.version;
+              badgeColor = vdata.color;
```
Comment (espadrine) [inline server.js]: Can you align `case` with `switch`, and use 2-space indentation, please?

File: server.js [hunk 2]
```diff
@@ -765,12 +766,44 @@ cache(function(data, match, sendBadge, request) {
...
+                          if (unstableVersion === null || (unstableVersion !== null && versionCompare(version, unstableVersion) > 0)) {
+                              unstableVersion = version;
+                          }
+                      }
+                  }
+
+                  badgeText = unstableVersion;
+                  badgeColor = 'orange';
+              }
+          break;
```
Comment (espadrine) [inline server.js]: Can you indent this break, please?

File: server.js [hunk 3]
```diff
@@ -765,12 +766,44 @@ cache(function(data, match, sendBadge, request) {
...
+      var badgeText = null;
+      var badgeColor = null;
+
+      switch (info) {
+          case 'v':
+              badgeText = vdata.version;
+              badgeColor = vdata.color;
+              break;
+          case 'vpre':
+              if (versions[0] !== undefined) {
```
Comment (espadrine) [inline server.js]: What is this condition taking care of?

File: server.js [hunk 4]
```diff
@@ -765,12 +766,44 @@ cache(function(data, match, sendBadge, request) {
...
+
+                  var unstableVersion = null;
+                  for (var i = 0; i < versions.length; i++) {
+                      var version = versions[i];
+
+                      if (version.charAt(0) !== 'v') {
```
Comment (espadrine) [inline server.js]: What if the version does start with a `v`? Do we never match a version at all?

As a side-note, `version[0] !== 'v'` sounds like what you mean; `charAt()` returns an integer, which will never match a string.

File: server.js [hunk 5]
```diff
@@ -765,12 +766,42 @@ cache(function(data, match, sendBadge, request) {
...
-      badgeData.text[1] = vdata.version;
-      badgeData.colorscheme = vdata.color;
+      var stableVersion = latestVersion(versions);
+      var vdata = versionColor(stableVersion);
+
+      var badgeText = null;
+      var badgeColor = null;
+
+      switch (info) {
+        case 'v':
```
Comment (espadrine) [inline server.js]: Can you align `case` with `switch` please?

File: server.js [hunk 6]
```diff
@@ -765,12 +766,42 @@ cache(function(data, match, sendBadge, request) {
...
+          badgeText = vdata.version;
+          badgeColor = vdata.color;
+          break;
+        case 'vpre':
+          if (versions[0] !== undefined) {
```
Comment (espadrine) [inline server.js]: What is this condition taking care of?

File: server.js [hunk 7]
```diff
@@ -765,12 +766,42 @@ cache(function(data, match, sendBadge, request) {
...
+            var unstableVersion = null;
+            for (var i = 0; i < versions.length; i++) {
+              var version = versions[i];
+              if (version.charAt(0) !== 'v') {
```
Comment (espadrine) [inline server.js]: What if the version does start with a `v`? Do we never match a version at all?

As a side-note, `version[0] !== 'v'` sounds like what you mean; `charAt()` returns an integer, which will never match a string.

**Comment by espadrine** [comment]
Thanks a lot for your work!

About the version comparison you introduce; it obviously doesn't work similarly to npm's semver version comparison algorithm. Is there a known distinct algorithm in the PHP community?

File: server.js (hunk 3 above)
Comment (WouterDS) [inline server.js]: Well, I suppose line 771 - 772 can be moved inside of the vpre case if that's what you mean?

L783 is just a habbit, checking if it exists, but might be cleaner to change that to if (versions) or if (versions.length > 0).

Comment (espadrine) [inline server.js]: Ah, I see. Yes, checking the length will be clearer.

Comment (WouterDS) [inline server.js]: Or actually we can maybe just get rid of the check on L783.

File: server.js (hunk 4 above)
Comment (WouterDS) [inline server.js]: If there are no versions starting without prefix v, it's all considered a stable release.

Hmm?

	var version = "v1.2.3";
	
	version[0]; // "v"
	version[1]; // "1"
	version.charAt(0); // "v"
	version.charAt(1); // "1"

Or am I wrong?

File: server.js (hunk 5 above)
Comment (WouterDS) [inline server.js]: Done

File: server.js (hunk 6 above)
Comment (WouterDS) [inline server.js]: Sorry, same question as https://github.com/badges/shields/pull/366#discussion-diff-24460127 ?

File: server.js (hunk 7 above)
Comment (WouterDS) [inline server.js]: Same question as https://github.com/badges/shields/pull/366#discussion-diff-24460294 ?

**Comment by WouterDS** [comment]
Thanks :).

I fixed the issues you pointed out, it should be ok now.

The [version comparison function](http://phpjs.org/functions/version_compare/) is borrowed from the [php.js project](https://github.com/kvz/phpjs). I'm not a big expert as well, but the method gives back the expected results in tests so I just blindly copied it.

**Comment by espadrine** [comment]
Thanks a lot!

**Comment by WouterDS** [comment]
Not sure what's live now on shields.io, but the vpre version for eg symfony/symfony, or laravel/framework is just wrong. And those are 2 of the biggest and most popular PHP repo's.

**Comment by espadrine** [comment]
Can you give me more information?

There was a recent change in the version comparison algorithm due to complaints [here](https://github.com/badges/shields/issues/319#issuecomment-74238688).

The change includes using aliases when available. I assumed that `version.extra['branch-alias']` mapped a branch alias corresponding to a specific branch to a version we can use. However, symfony has a dev-propel1-removal alias which doesn't contain the corresponding version in the `branch-alias` field.

That seems to go contrary to what is explained [here](https://getcomposer.org/doc/articles/aliases.md):

> The alias must be a comparable dev version, and the branch-alias must be present on the branch that it references.

I suppose I could drop support for aliases and filter out all “versions” starting with `dev`.

**Comment by espadrine** [comment]
@WouterDS Please tell me if you feel like this change, which is live, is satisfactory.

**Comment by WouterDS** [comment]
@espadrine What's live now is good :)!

**Pull request merged**